packages/nx/src/project-graph/utils/project-configuration-utils.ts

```diff
@@ -31,53 +36,148 @@ export function mergeProjectConfigurationIntoRootMap(
+      projectConfigSourceMap?.set(
+        `${project.root}.namedInputs.${namedInput}`,
+        sourceInformation
+      );
+    }
   }
 
-  if (project.targets && matchingProject.targets) {
+  if (project.targets) {
     updatedProjectConfiguration.targets = {
-      ...matchingProject.targets,
+      ...(matchingProject?.targets ?? {}),
       ...project.targets,
     };
-    for (const target in matchingProject.targets) {
-      if (target in matchingProject.targets && target in project.targets) {
-        // If the target is defined in both places, merge the options
-        // Project contains the new info, so it has higher priority.
-        // Options from matchingProject are used as defaults.
-        updatedProjectConfiguration.targets[target] = mergeTargetConfigurations(
-          project.targets[target],
-          matchingProject.targets[target]
-        );
+
+    if (matchingProject.targets) {
+      for (const target in matchingProject.targets) {
+        if (target in matchingProject.targets && target in project.targets) {
+          // If the target is defined in both places, merge the options
+          // Project contains the new info, so it has higher priority.
+          // Options from matchingProject are used as defaults.
+          updatedProjectConfiguration.targets[target] =
+            mergeTargetConfigurations(
+              project.targets[target],
+              matchingProject.targets[target],
+              projectConfigSourceMap,
+              `${project.root}.targets.${target}`
+            );
+        }
+      }
+    }
+
+    if (projectConfigSourceMap) {
+      for (const target in project.targets) {
+        const targetId = `${project.root}.targets.${target}`;
+        projectConfigSourceMap.set(targetId, sourceInformation);
+
+        for (const targetProperty in project.targets[target]) {
+          const targetPropertyId = `${targetId}.${targetProperty}`;
+          projectConfigSourceMap.set(targetPropertyId, sourceInformation);
+
+          if (targetProperty === 'options') {
+            for (const option in project.targets[target].options) {
+              projectConfigSourceMap.set(
+                `${targetPropertyId}.${option}`,
+                sourceInformation
+              );
+            }
+          }
+
+          if (targetProperty === 'configurations') {
+            for (const configuration in project.targets[target]
+              .configurations) {
+              projectConfigSourceMap.set(
+                `${targetPropertyId}.${configuration}`,
+                sourceInformation
+              );
+              for (const configurationProperty in project.targets[target]
+                .configurations[configuration]) {
+                projectConfigSourceMap.set(
+                  `${targetPropertyId}.${configuration}.${configurationProperty}`,
+                  sourceInformation
+                );
+              }
+            }
+          }
+        }
```

Comment:
I feel like this part should be done during the `mergeTargetConfigurations` and `mergeConfigurations` calls. As is, we introduce multiple places that have to know how targets are merged (configurations one level deep and such). I know this doesn't actually handle the merge, but it has to know about the inner workings of the above functions. Ideally, they are black boxes that handle the merge operation and recording info in the source map.

I also feel like we already have most of the information in those functions to do this, especially with passing in the keys as you have for when the targets aren't compatible.